[draft] Reduce constants duplication

m-bronnikov opened the pull request:
This draft reduces constants duplication via providing special GraphBuilderSource.

Signed-off-by: Maksim Bronnikov <max120199@gmail.com>

-----------------

For: #8042
Another version of: #8043

m-bronnikov commented:
## Summary of changes

Let me please explain how this draft allows to not copy constant values.

### Change GraphBuilderSource interface in [luci/import[

Currently, Importer creates constant values using internal `create_circleconst` function when build of other `CircleNode`s delegated to `GrpahBuilderSource` entity. I think it's more natural to delegate build of constant nodes to `GraphBuilderSource` too. 

https://github.com/Samsung/ONE/blob/9676cceedcc1fb146287bf7abebc21a10f056045/compiler/luci/import/include/luci/Import/GraphBuilderRegistry.h#L36

https://github.com/Samsung/ONE/blob/9676cceedcc1fb146287bf7abebc21a10f056045/compiler/luci/import/src/Importer.cpp#L128

Since `Importer` allows to pass custom `GraphBuilderSource` as ctor argument, `Importer`'s user can define any behavior to this method. We will use this ability to provide our own constant node, which will store pointer to constants from model buffer.

### Add [luci-micro-extension] module

This module provides helper methods to use luci::Importer without constant duplication using custom `GraphBuilderRegistry`:

https://github.com/Samsung/ONE/blob/9676cceedcc1fb146287bf7abebc21a10f056045/compiler/luci-micro-extension/import/include/luci_micro/Import/MicroGraphBuilderRegistry.h#L28-L31

Main features:

1. Introduce function to import constants as `CircleCustom` op (instead `CircleConst`, which requires duplication). This node contain pointer to constant buffer with buffer's size.
https://github.com/Samsung/ONE/blob/9676cceedcc1fb146287bf7abebc21a10f056045/compiler/luci-micro-extension/import/include/luci_micro/Import/Nodes/CircleConstProxy.h#L31

 https://github.com/Samsung/ONE/blob/9676cceedcc1fb146287bf7abebc21a10f056045/compiler/luci-micro-extension/import/src/Nodes/CircleConstProxy.cpp#L87

https://github.com/Samsung/ONE/blob/9676cceedcc1fb146287bf7abebc21a10f056045/compiler/luci-micro-extension/import/src/Nodes/CircleConstProxy.cpp#L102-L110

2. Use this function inside of new `GraphBuilderRegistry`
https://github.com/Samsung/ONE/blob/9676cceedcc1fb146287bf7abebc21a10f056045/compiler/luci-micro-extension/import/src/MicroGraphBuilderRegistry.cpp#L169-L173

### Add support of new op to [luci-interpreter]

This part of changes just adds support of "CircleConstProxy" `CircleCustom` node to luci-interpreter.

### Simple verification using [model-import-verifier]

This module needed for simple proof of correctness which similar to model-import-verifier from #8043

#### Note

This draft done for approach demonstration. Current changes are much easier to understand, but if this approach looks good for reviewers, I think better to follow suggested by @binarman design https://github.com/Samsung/ONE/pull/8043#issuecomment-982959618.

+cc @binarman @seanshpark

seanshpark commented:
Thanks for preparing this draft. I think it's better than previous one.
I agree on overriding and extending the interface.

I've some questions and comments;

(1) I can't catch the idea why `CircleCustom` is mentioned here.
> Introduce function to import constants as CircleCustom op

(2) `luci-micro-extension` is introduced cause the main purpose is to enable `luci-micro` with no copy constant?

(3) I think `create_const` inside `GraphBuilderSource` somewhat odd. Main purpose of this class was for graph building but `create_const` is one specific node (I understand this is the main issue)

(4) About Copy codes in `MicroGraphBuilderRegistry`, `CIRCLE_NODE(ABS, CircleAbsGraphBuilder);` and others. Do these need to sync with existing code? I think they can be out-of-sync but I'd like to be sure. I can't verify nor test codes of `luci-extension` so it would be hard for me to work on this module.

(5) Why is the name `create_circleconst_proxy` have `proxy` ?

(6) About the name `create_const`. It was named like this cause it was local without much thoughts, but now this seems to be used as an interface. I don't think `create_const` is a good name as an interface. Currently I can't think of any good name, not having much understanding of this draft as of now and not much time to think. I'll propose when a good name popsup, or I think you can suggest a good name.

I'll add more when I find more things...

m-bronnikov commented:
> Thanks for preparing this draft. I think it's better than previous one. I agree on overriding and extending the interface.
> 
> I've some questions and comments;
> 
> (1) I can't catch the idea why `CircleCustom` is mentioned here.
> 
> > Introduce function to import constants as CircleCustom op
> 
> (2) `luci-micro-extension` is introduced cause the main purpose is to enable `luci-micro` with no copy constant?
> 
> (3) I think `create_const` inside `GraphBuilderSource` somewhat odd. Main purpose of this class was for graph building but `create_const` is one specific node (I understand this is the main issue)
> 
> (4) About Copy codes in `MicroGraphBuilderRegistry`, `CIRCLE_NODE(ABS, CircleAbsGraphBuilder);` and others. Do these need to sync with existing code? I think they can be out-of-sync but I'd like to be sure. I can't verify nor test codes of `luci-extension` so it would be hard for me to work on this module.
> 
> (5) Why is the name `create_circleconst_proxy` have `proxy` ?
> 
> (6) About the name `create_const`. It was named like this cause it was local without much thoughts, but now this seems to be used as an interface. I don't think `create_const` is a good name as an interface. Currently I can't think of any good name, not having much understanding of this draft as of now and not much time to think. I'll propose when a good name popsup, or I think you can suggest a good name.
> 
> I'll add more when I find more things...

Thank you for your review! Let me answer on your questions:

1. Main purpose of this draft, as of previous one, is to provide size and pointer of constant values placed in model's buffer. Prev draft introduced new `CircleNode` (named `CircleRemoteConst`) in luci, but this is MCU specific node and this draft tries to extract MCU specific features from `luci` to another place (`luci-micro-extension`). So, I implemented node similar to `CircleRemoteConst` in this draft via `CircleCustom` node.
2. `luci-micro-extension` introduced as module, which adds MCU specific functional to `luci` without `luci` modification. So, answer is yes.
3. Since `CircleConst` is one of the possible `grpah`'s nodes, I think it's natural to delegate constant's building to this class too. 
4. `CIRCLE_NODE(...)`s from `MicroGraphBuilderRegistry`'s ctor just copied from `luci::GraphBuilderRegistry`'s ctor. These classes differ by their `create_const` implementation. 
5. Last draft showed `CircleRemoteCosnt` and `CircleConstReference` is not good names for new added op. I decided to add "Proxy" preffix to new node in accordance with @binarman suggestion https://github.com/Samsung/ONE/pull/7821#discussion_r732013672. `create_circleconst_proxy` named because creates `CircleConstProxy` node, but I think we could change this name.
6. As I mentioned before I implemented `circle_const` method in draft to make changes easier to understand, but I would like to follow design proposed in https://github.com/Samsung/ONE/pull/8043#issuecomment-982959618.

seanshpark commented:
OK, thanks share your thoughts :)

> 1. ... but via CircleCustom node.

OK, that is this `CircleCustom` node is actually CircleConst but without copy.
I think this will make the modules (maybe luci-interpreter and others) difficult with lots of `if-else` to check if it is CircleConst or not.
If you look into `luci/lang` where it is derived from _loco_ as a dialect, you can define another dialect for luci-micro. There will be some amount of codes to provide but will make smooth handling of IRs for micro. This is your choice, I'm just providing a way of implementation.

> Since CircleConst is one of the possible grpah's nodes, I think it's natural to delegate constant's building to this class too.

I'm still not convinced to add a methon into `GraphBuilderSource` class. It would be better to add a general interface name, such as without `const`.

m-bronnikov commented:
> OK, that is this `CircleCustom` node is actually CircleConst but without copy. I think this will make the modules (maybe luci-interpreter and others) difficult with lots of `if-else` to check if it is CircleConst or not. If you look into `luci/lang` where it is derived from _loco_ as a dialect, you can define another dialect for luci-micro. There will be some amount of codes to provide but will make smooth handling of IRs for micro. This is your choice, I'm just providing a way of implementation.

Personally, I do not want to introduce new dialect because I don't know how many extra nodes like introduced `CircleConstProxy` we need, but amount of changes (especially for `luci-interpeter`) looks huge to me. I think we could return to this approach later if luci_micro will provide more `CustomNode`s. 
 
> I'm still not convinced to add a methon into `GraphBuilderSource` class. It would be better to add a general interface name, such as without `const`.

I will try to rewrite this code now. =)

binarman commented on compiler/luci/import/include/luci/Import/GraphBuilderRegistry.h:
```diff
@@ -73,11 +90,17 @@ class GraphBuilderRegistry final : public GraphBuilderSource
 
 private:
   std::map<const circle::BuiltinOperator, std::unique_ptr<GraphBuilderBase>> _builder_map;
+  std::vector<std::unique_ptr<TensorBuilder>> _tensor_builders;
```
nitpick: you can use 
``` c++
std::array<std::unique_ptr<TensorBuilder>, TensorBuilderType::COUNT> _tensor_builder = {};
```
 or
 ``` c++
 std::unique_ptr<TensorBuilder> _tensor_builder[TensorBuilderType::COUNT] = {}
 ```

m-bronnikov replied:
Personally, I prefer first option. Done!

binarman commented on compiler/luci/import/include/luci/Import/TensorBuilder.h:
```diff
@@ -0,0 +1,61 @@
+/**
+ * @brief Creates nodes from given Tensor and context
+ */
+class TensorBuilder
```
I have concerns related to the name of this entity, because it builds nodes (not tensors :slightly_frowning_face: )

m-bronnikov replied:
I feel this so. However, I have not thought of a better name. I'm happy to hear any suggestions about naming!

m-bronnikov replied:
How about `SpecialNodeBuilder`?

m-bronnikov replied:
Renamed as `SpecialNodeBuilder` in current revision.

binarman replied:
Another thing that came to mind is `EndPointNodeBuilder`

binarman commented on compiler/luci-interpreter/src/import/Nodes/CircleCustomConst.cpp:
```diff
@@ -0,0 +1,124 @@
+  // create CircleCustomConst
+  auto custom_node = graph->nodes()->create<CircleCustom>(0, 1);
+  {
+    custom_node->custom_code("CircleCustomConst");
```
maybe be more specific?
like `CircleReferencingConst`

m-bronnikov replied:
Done. Renamed!

binarman commented on compiler/model-import-verifier/src/EvalDriver.cpp:
```diff
@@ -0,0 +1,161 @@
+  std::vector<model_dtype> output_data_2;
+  {
+    // default builder source not allows using constants from model's buffer
+    auto default_source = &luci::GraphBuilderRegistry::get();
```
Thank you for usage example!
I think you can highlight this part in Notes, so it will be easier to find it in this PR (it is large :slightly_smiling_face: )

m-bronnikov replied:
As far I understand your suggestion I did that.

binarman commented on compiler/luci/import/src/GraphBuilderRegistry.cpp:
```diff
@@ -161,6 +161,14 @@ GraphBuilderRegistry::GraphBuilderRegistry()
   // BuiltinOperator_HARD_SWISH = 117,
   // BuiltinOperator_DENSIFY = 124,
+
+#define CIRCLE_VNODE(OPCODE, CLASS) add(std::make_unique<CLASS>())
```
Do we need `OPCODE`?

m-bronnikov replied:
No, we don't need, but I made this definition in accordance with definition for not virtual circle nodes (see above in this file).

seanshpark replied:
I think this is where lookup source is added.
We need a comment to express this is where tensor type(?) builders (CircleConst) are build.

binarman commented:
Overall looks good to me, but there are some names and arguments that bothers me a little =)

m-bronnikov commented on compiler/luci/import/src/Nodes/CircleConst.cpp:
```diff
@@ -149,8 +134,6 @@ CircleConst *create_circleconst(GraphBuilderContext *context, int32_t tensor_ind
   auto const_node = graph->nodes()->create<CircleConst>();
   copy_tensor_attributes(const_tensor, const_node);
   const_node->shape_status(luci::ShapeStatus::VALID);
-  INFO(l) << "[luci] NodeFinder const_node(" << tensor_index << ") -> " << const_node << " "
```
Personally, I would like to move logger to caller (moved to Importer.cpp). Reason: do not duplicate logger code in each builder. But I can return logger to builder if you want.

m-bronnikov replied:
I returned logger back! =)

seanshpark replied:
I think logger code was put back to another place.

m-bronnikov commented on compiler/luci/import/src/Nodes/CircleConst.cpp:
```diff
@@ -107,9 +91,10 @@ void copy_data<loco::DataType::STRING>(const VectorWrapper<uint8_t> &raw_data,
 namespace luci
 {
 
-CircleConst *create_circleconst(GraphBuilderContext *context, int32_t tensor_index)
```
Ahh, you are right, but I made context as second arg because `GraphBuilder::build` makes context as second arg:

```cpp
CircleNode *build(const circle::OperatorT &op, GraphBuilderContext *context) const final;
```

m-bronnikov commented:
### Summary of changes

@seanshpark Let me describe what changed from your last review.

1. Custom node `CircleConstProxy` renamed to `CircleReferencingConst` according to https://github.com/Samsung/ONE/pull/8190#discussion_r774106503 suggestion.

 > (3) I think create_const inside GraphBuilderSource somewhat odd. Main purpose of this class was for graph building but create_const is one specific node (I understand this is the main issue)
> (6) About the name create_const. It was named like this cause it was local without much thoughts, but now this seems to be used as an interface. I don't think create_const is a good name as an interface. Currently I can't think of any good name, not having much understanding of this draft as of now and not much time to think. I'll propose when a good name popsup, or I think you can suggest a good name.
2. Introduced new type of builders: `SpecialNodeBuilder`. This inheritors of this builder creates nodes from: Input Tensors, Output Tensors, Tensors with Buffers (constants).
```cpp
/**
 * @brief Tensor types which requires separated node
 */
enum class SpecialNodeBuilderType
{
  INPUT,
  BUFFER,
  OUTPUT,
  // TODO Extend this struct here if needed to add new type of SpecialNodeBuilder
  COUNT // Count of different types
};

/**
 * @brief Creates nodes from given Tensor and context
 */
class SpecialNodeBuilder
{
public:
  virtual CircleNode *build(TensorIndex tensor_idx, GraphBuilderContext *context) const = 0;
  virtual SpecialNodeBuilderType builder_type() const = 0;
};

/**
 * @brief Placeholder for builders of tensors with different types
 */
template <SpecialNodeBuilderType Type> class SpecialNodeBuilderImpl : public SpecialNodeBuilder
{
public:
  SpecialNodeBuilderType builder_type() const final { return Type; }
};
```
With this class we can extend interface of GrpahBuilderSource with new lookup method:
```diff
@@ -32,6 +33,11 @@ struct GraphBuilderSource
    * @brief Returns registered GraphBuilder pointer for operator (nullptr if not present)
    */
   virtual const GraphBuilderBase *lookup(const circle::BuiltinOperator &op) const = 0;
+
+  /**
+   * @brief Returns registered SpecialNodeBuilder pointer for type (nullptr if not present)
+   */
+  virtual const SpecialNodeBuilder *lookup(SpecialNodeBuilderType type) const = 0;
 };
};
```
And then:
- Get ability to redefine builders (make custom builder for constants in our case).
- Simplify logic of inputs, outputs and constants creation in `convert_graph()` in *Importer.cpp*

> I can't verify nor test codes of luci-extension so it would be hard for me to work on this module.

3. Since `CircleReferencingConst` is node introduced for usage in `luci-interpreter` and this custom node handles only in this module, I decided to move new builder for this op to `luci-interpreter/import`.

-------------

PTAL =)

m-bronnikov commented:
@seanshpark Could you please take a look?)

seanshpark commented on compiler/luci/import/include/luci/Import/GraphBuilderRegistry.h:
```diff
@@ -32,6 +33,11 @@ struct GraphBuilderSource
+  /**
+   * @brief Returns registered SpecialNodeBuilder pointer for type (nullptr if not present)
+   */
+  virtual const SpecialNodeBuilder *lookup(SpecialNodeBuilderType type) const = 0;
```
About word `Special`. I don't think using `Special` that is ambigious in purpose is good to use. Please use specific purpose. If this is to provide a tensor with data, please provide a word that relfects that purpose.

seanshpark replied:
I like the idea of introducing `~~~NodeBuilder`.

seanshpark replied:
To follow existing `GraphBuilderBase` name, I think `NodeBuilderBase` would be OK.

seanshpark commented:
I don't understand why Input and Output nodes are changed also.
If the changes aren't necessary, please leave as it is.

seanshpark commented on compiler/luci/import/include/luci/Import/SpecialNodeBuilder.h:
```diff
@@ -0,0 +1,61 @@
+/**
+ * @brief Placeholder for builders of tensors with different types
+ */
+template <SpecialNodeBuilderType Type> class SpecialNodeBuilderImpl : public SpecialNodeBuilder
```
Can there be other name not using `Impl`?
This looks javaish and I'd prefer using name that express itself for it's purpose.

seanshpark commented:
If possible how about conveting this draft ot `DRAFT` ?

![image](https://user-images.githubusercontent.com/4616940/147523962-e6845606-4783-4492-92de-2af75ddc0a08.png)

m-bronnikov commented:
> If possible how about conveting this PR to `DRAFT` ?

Yes, it possible =)

I didn't do this in order to pass CI tests on introduced changes.

m-bronnikov commented:
> I don't understand why `Input` and `Output` nodes are changed also. If the changes aren't necessary, please leave as it is.

The reason why I changed `Input`/`Output` is following:

There is 3 types of tensors, which requires special `CircleNode`, but not has corresponding operator:
- Graph's input tensors
- Graph's output tensors
- Tensors with bound buffers

This draft introduces `SpecialNodeBuilder` (I agree better to change name), which generalizes creation of nodes for this tensors and introduces 3 different types of builders:

https://github.com/Samsung/ONE/blob/e7cca5e88386ae210a36cc4fcb13f7e7662d2ae8/compiler/luci/import/include/luci/Import/SpecialNodeBuilder.h#L31-L38

So, I decided to introduce 3 builders for `Input`, `Const` and `Output` nodes, and, then, simplify logic of `convert_graph` method by delegating node building to `SpecialNodeBuilder`'s entities.

**However,** you are right, `Input`/`Output` changes are not related with constant duplication reduction and may be removed from this draft. But this changes related with introducing new `~~~NodeBuilder`.

I will rebase and remove this changes.

m-bronnikov commented:
@seanshpark I introduced changes to draft in accordance with your suggestions. PTAL =)

> To follow existing `GraphBuilderBase` name, I think `NodeBuilderBase` would be OK.

Renamed to `NodeBuilderBase` as you suggested. Also `SpecialNodeBuilderType` renamed to `NodeBuilderType` and `SpecialNodeBuilder.h` renamed to `NodeBuilder.h`

> Can there be other name not using `Impl`? This looks javaish and I'd prefer using name that express itself for it's purpose.

Renamed to `TypedNodeBuilder`. What do you think about this name.

> I don't understand why Input and Output nodes are changed also. If the changes aren't necessary, please leave as it is.

Removed changes related to Input/Output.

Also, introduced some small fixes as `nullptr` dereference check applied to new builder.

seanshpark commented:
> I didn't do this in order to pass CI tests on introduced changes.

CI will run with DRAFT PR also. 
CI will NOT run when `PR/NO TEST` label is added.

seanshpark commented:
> There is 3 types of tensors, which requires special CircleNode, but not has corresponding operator:

For this purpose, I think, there are also other virtual nodes. They were created for "Tensors" but don't have corresponding operator.

Still I don't understand why they are treated in "SpecialNodeBuilder" class.

seanshpark commented on compiler/luci-interpreter/include/luci_interpreter/import/Nodes/CircleReferencingConst.h:
```diff
@@ -0,0 +1,39 @@
+ * limitations under the License.
+ */
+
+#ifndef __LUCI_INTERPRETER_IMPORT_OP_CIRCLE_CUSTOM_CONST_H__
```
Maybe this needs change to follow file name.

m-bronnikov replied:
Done!

seanshpark commented on compiler/luci-interpreter/include/luci_interpreter/import/Nodes/CircleReferencingConst.h:
```diff
@@ -0,0 +1,39 @@
+/**
+ * @brief Builder creates CircleCustom node with pointer to constants data from Tensor with buffer.
+ */
+class CircleReferencingConstSpecialNodeBuilder : public TypedNodeBuilder<NodeBuilderType::BUFFER>
```
(just my opinion as I'm not a maintainer of luci interpreter) maybe better to not use `Special` word but use a word for express its purpose.

m-bronnikov replied:
Fixed, thank you!

seanshpark commented on compiler/luci/import/include/luci/Import/GraphBuilderRegistry.h:
```diff
@@ -61,6 +67,17 @@ class GraphBuilderRegistry final : public GraphBuilderSource
+   */
+  const NodeBuilderBase *lookup(NodeBuilderType type) const final
+  {
+    if (_tensor_builders.at(uint32_t(type)) == nullptr)
```
I don't think `at(uint32_t(type))` is safe. maybe use `map` ?

m-bronnikov replied:
`std::map` is applied instead `std::array`, `at(uint32_t(type))` usage removed.

seanshpark commented on compiler/luci/import/include/luci/Import/GraphBuilderRegistry.h:
```diff
@@ -73,11 +90,18 @@ class GraphBuilderRegistry final : public GraphBuilderSource
 
+  void add(std::unique_ptr<NodeBuilderBase> &&builder)
+  {
+    _tensor_builders.at(uint32_t(builder->builder_type())) = std::move(builder);
```
ditto; `uint32_t(builder->builder_type()` doesn't seem to be safe.

seanshpark commented on compiler/luci/import/include/luci/Import/NodeBuilder.h:
```diff
@@ -0,0 +1,61 @@
+ * limitations under the License.
+ */
+
+#ifndef __LUCI_IMPORT_SPECIAL_NODE_BUILDER_H__
```
Maybe this need change to follow the file name.

m-bronnikov replied:
Fixed (changed)!

seanshpark commented on compiler/luci/import/include/luci/Import/NodeBuilder.h:
```diff
@@ -0,0 +1,61 @@
+ */
+enum class NodeBuilderType
+{
+  INPUT,
+  BUFFER,
+  OUTPUT,
```
I still don't understand/agree on treating input/output with `NodeBuilderType`

seanshpark commented on compiler/luci/import/src/Importer.cpp:
```diff
@@ -122,12 +122,19 @@ void convert_graph(const luci::GraphBuilderSource &source, luci::CircleReader &r
   // Create CircleConst nodes for constant tensors.
+  // Create CircleNodes for constant tensors.
   // NOTE Origin is intentionally not provided for constants.
+  auto const_builder = source.lookup(luci::NodeBuilderType::BUFFER);
+  if (not const_builder)
+    throw oops::UserExn("Not supported", "tensor with buffer builder");
```
Could you explain more about why this change is introduced?

m-bronnikov replied:
This change is introduced because `source.lookup()` returns nullptr in case is builder is not found. So, we need to check `nullptr` before using returned builder.

https://github.com/Samsung/ONE/blob/daea9fdc9f8a6338d14d2be88c14f7187256be93/compiler/luci/import/include/luci/Import/GraphBuilderRegistry.h#L37-L40

We also make this check with builder for operators:

https://github.com/Samsung/ONE/blob/daea9fdc9f8a6338d14d2be88c14f7187256be93/compiler/luci/import/src/Importer.cpp#L151
https://github.com/Samsung/ONE/blob/daea9fdc9f8a6338d14d2be88c14f7187256be93/compiler/luci/import/src/Importer.cpp#L170-L173

seanshpark commented on compiler/luci/import/src/Importer.cpp:
```diff
@@ -122,12 +122,19 @@ void convert_graph(const luci::GraphBuilderSource &source, luci::CircleReader &r
   for (uint32_t i = 0; i < tensors.size(); ++i)
   {
-    luci::CircleConst *const_node = luci::create_circleconst(&gb_context, i);
+    auto *const_node = const_builder->build(i, &gb_context);
```
Q) why changed to auto?

m-bronnikov replied:
We can't use `luci::CircleConst` here because `build()` method returns `luci::CircleNode`. As another option we can change to `luci::CircleNode` instead `auto` but `auto` looks better for me.

seanshpark commented on compiler/model-import-verifier/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+target_link_libraries(model_eval_verifier PRIVATE luci_interpreter)
+target_link_libraries(model_eval_verifier PRIVATE safemain)
+
+install(TARGETS model_eval_verifier DESTINATION bin)
```
(optional) We used `tflite` or `circle` or event `caffe` for specific type of model not general `model` that means, one may think this `model-import-verifier` can be any model import verifier, which I don't think you intend to do. Or do you?

m-bronnikov replied:
I don't plan to introduce `model-import-verifier` module. I provide this only as example and simple proof  for this draft.

seanshpark replied:
You mean "land"? You already "introduced"(?) the module through this draft.
I think landing this module will actually "verify" if you revise this with gtest.
As of current draft, I do not see any tests for new codes.

m-bronnikov replied:
I agree with you: better to rewrite with gtest, but let me please add this change later, when code of other changes (code from `luci` and `luci-interpreter`) will looks OK for reviewers.

seanshpark commented:
> Renamed to TypedNodeBuilder. What do you think about this name?

Looks OK :) thank you!

seanshpark commented:
> Also, introduced some small fixes as nullptr dereference check applied to new builder.

Which change may I ask?

seanshpark commented on compiler/luci-interpreter/include/luci_interpreter/import/GraphBuilderRegistry.h:
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
```
I think it would be better to revise all `2021` to `2022`.
Happy new year!

m-bronnikov replied:
Happy New Year! =)

m-bronnikov replied:
Replaced with 2022 year!

m-bronnikov commented:
> > Also, introduced some small fixes as nullptr dereference check applied to new builder.
> 
> Which change may I ask?

Ahh, you already found this change: https://github.com/Samsung/ONE/pull/8190#discussion_r777261489

seanshpark commented on compiler/luci/import/src/Importer.cpp:
```diff
@@ -122,12 +122,19 @@ void convert_graph(const luci::GraphBuilderSource &source, luci::CircleReader &r
   // Create CircleConst nodes for constant tensors.
+  // Create CircleNodes for constant tensors.
   // NOTE Origin is intentionally not provided for constants.
+  auto const_builder = source.lookup(luci::NodeBuilderType::BUFFER);
```
Can you please point where builder is registered with `BUFFER`?

m-bronnikov replied:
Builder is registred in same way as builder regiters for other nodes:

https://github.com/Samsung/ONE/blob/daea9fdc9f8a6338d14d2be88c14f7187256be93/compiler/luci/import/src/GraphBuilderRegistry.cpp#L165-L170

seanshpark commented on compiler/luci/import/src/Importer.cpp:
```diff
@@ -122,12 +122,19 @@ void convert_graph(const luci::GraphBuilderSource &source, luci::CircleReader &r
   }
 
   // Create CircleConst nodes for constant tensors.
+  // Create CircleNodes for constant tensors.
```
These two comments conflict.

m-bronnikov replied:
Removed `  // Create CircleConst nodes for constant tensors.` comment!

seanshpark commented on compiler/luci/import/include/luci/Import/GraphBuilderRegistry.h:
```diff
@@ -32,6 +33,11 @@ struct GraphBuilderSource
+  /**
+   * @brief Returns registered NodeBuilderBase pointer for type (nullptr if not present)
+   */
+  virtual const NodeBuilderBase *lookup(NodeBuilderType type) const = 0;
```
Can we use `const NodeBuilderType type` ?

m-bronnikov replied:
Yes, we can!

m-bronnikov replied:
Done!

seanshpark commented on compiler/luci/import/src/GraphBuilderRegistry.cpp:
```diff
@@ -161,6 +161,13 @@ GraphBuilderRegistry::GraphBuilderRegistry()
+#define CIRCLE_VNODE(OPCODE, CLASS) add(std::make_unique<CLASS>())
+
+  // TODO add CIRCLEINPUT and CIRCLEOUTPUT builder
+  CIRCLE_VNODE(CIRCLECONST, CircleConstNodeBuilderBase);
```
Why is it `CircleConstNodeBuilderBase`?

seanshpark replied:
I thought name `Base` is to for base class and you used for implementation.
Plese don't confuse using names.

m-bronnikov replied:
Ahh, better to change to `CircleConstNodeBuilder`. `Base` added accidentally, sorry.

m-bronnikov replied:
Fixed!

m-bronnikov commented:
> > There is 3 types of tensors, which requires special CircleNode, but not has corresponding operator:
> 
> For this purpose, I think, there are also other virtual nodes. They were created for "Tensors" but don't have corresponding operator.
> 
> Still I don't understand why they are treated in "SpecialNodeBuilder" class.

When I proposed this change, I was thinking that in the convert_graph function we separately handle the creation of CircleInput, CircleConst and CirceOutput, and it seemed to me that their creation can be delegated to SpecialNodeBuilder. This could give the following advantages:
- `convert_graph()` more simple and easy to read.
- For each node some builder exist.
- This builders may be registered and redefined in `GraphBuilderRegistry`. It gives more flexibility for user.

seanshpark commented:
> This could give the following advantages:

You got a point. Now I understand and agree on your plan!

m-bronnikov commented:
> > This could give the following advantages:
> 
> You got a point. Now I understand and agree on your plan!

Thank you! I would follow this plan, but I removed changes related with `Input`/`Output` from this draft for now in order to reduce code changes. I think we can discuss details of this plan implementation for these nodes later as another issue.

m-bronnikov commented:
@seanshpark I fixed your review comments except comment https://github.com/Samsung/ONE/pull/8190#discussion_r777747392 related with `model-import-verifier` (reason: https://github.com/Samsung/ONE/pull/8190#discussion_r781110140). Could you please take a look again? =)

seanshpark commented:
> Could you please take a look again? =)

Thanks for the changes. Current PR looks OK.

Q) just to be sure, using `CircleReferencingConst` in luci_interpreter is not active unless `source_without_constant_copying` is called?

seanshpark commented:
@nnfw-bot test nncc-debug

m-bronnikov commented:
> Q) just to be sure, using `CircleReferencingConst` in luci_interpreter is not active unless `source_without_constant_copying` is called?

Yes, If user not calls `source_without_constant_copying` default `GraphBuilderRegistry` will in use, which know nothing about `CircleReferencingConst` and use `CircleConst` node.

m-bronnikov commented:
### What changed from last review?

1. Small changes in luci-interpeter move CircleReferencingConst.h to `src` and, therefore, hide `CircleReferencingConstNodeBuilder` from external user (now user can use *not copy constants* feature only via `source_without_constant_copying` function).
2. Remove `model-eval-verifier` and introduce `embedded-import-value-test` module, which provides value testing of imported without constant duplication models from *common-artifacts*. I added this module because I failed to rewrite `model-eval-verifier` with gtest in accordance with https://github.com/Samsung/ONE/pull/8190#discussion_r781110140. PTAL https://github.com/Samsung/ONE/blob/73aeb8d1367d0aa8e3781f152bb6452e07c4365e/compiler/embedded-import-value-test/README.md for details of introduced check.

@seanshpark Could you please take a look and share your opinion?

m-bronnikov commented on compiler/embedded-import-value-test/test.lst:
```diff
@@ -0,0 +1,190 @@
+
+# Simple Network test
+addeval(Part_While_000)
+addeval(Part_While_001)
```
List copied from https://github.com/Samsung/ONE/blob/73aeb8d1367d0aa8e3781f152bb6452e07c4365e/compiler/luci-value-test/test.lst.

seanshpark replied:
Why are most of them comment with `#` ?

m-bronnikov replied:
I just copied this file from `luci-value-test`, but as far as I know this tests can't be proceeded because not supported by `luci-interpreter`. I tried to uncomment this lines locally and obtain fail of following tests:

<details>
<summary> Failed tests </summary>

```
FAILED
- Abs_000
- AddN_000
- ArgMin_000
- ArgMin_001
- ArgMin_002
- ArgMin_003
- ArgMin_U8_000
- ArgMin_U8_001
- ArgMin_U8_002
- ArgMin_U8_003
- BatchMatMul_000
- BatchMatMulV2_000
- BatchMatMulV2_001
- Ceil_000
- Cos_000
- ExpandDims_000
- ExpandDims_001
- ExpandDims_002
- ExpandDims_003
- Fill_000
- Fill_001
- FloorMod_000
- FloorMod_001
- FullyConnected_U8_000
- GatherNd_000
- L2Pool2D_U8_000
- Log_000
- MatMul_000
- MatrixDiag_000
- MatrixSetDiag_000
- Mul_U8_000
- OneHot_003
- Range_000
- Rank_000
- ReduceAny_000
- ReduceAny_001
- ReduceAny_002
- ReduceAny_003
- ReduceMax_000
- ReduceMin_000
- ReduceProd_000
- ReduceProd_001
- ReduceProd_002
- ReduceProd_003
- ReLUN1To1_000
- Reshape_003
- ReverseSequence_000
- ReverseV2_000
- Round_000
- ScatterNd_000
- SegmentSum_000
- Select_000
- Select_001
- Select_002
- SelectV2_000
- SelectV2_001
- SelectV2_002
- Shape_000
- Sin_000
- SparseToDense_000
- Sum_000
- Sum_001
- Tile_000
- Tile_U8_000
- TopKV2_000
- TopKV2_001
- Where_000
- Where_001
- While_003
- ZerosLike_000
```

</details>

Log of this tests reports corresponding models contains unsupported operators =(

```
++ /home/mbronnikov/Work/ONE/build/compiler/embedded-import-value-test/test_driver --model /home/mbronnikov/Work/ONE/build/compiler/common-artifacts/AddN_000.circle
ERROR: Unsupported operator: 2 ofm
```

m-bronnikov replied:
From other hand not all commented cases unsupported. After `luci-interpreter` improvement some models are supported (e.g. Cast, OneHot). So I think better to enable this tests here and in https://github.com/Samsung/ONE/blob/73aeb8d1367d0aa8e3781f152bb6452e07c4365e/compiler/luci-value-test/test.lst.

m-bronnikov commented:
@nnfw-bot test nncc-debug

m-bronnikov commented:
@nnfw-bot test nncc-release

m-bronnikov commented:
@nnfw-bot test android-runtime-builder

seanshpark commented:
> @seanshpark Could you please take a look and share your opinion?

Left a comment `Why are most of them comment with # ?`.
Overall code looks good to me :)

m-bronnikov commented:
Close as merged =)
